File: jsimple-oauth/src/main/java/jsimple/oauth/builder/api/DefaultOAuthApi10a.java
/*
 * Portions copyright (c) 2012-2014 Microsoft Mobile.  All Rights Reserved.
 *
 * This file is based on or incorporates material from the Scribe Java OAuth
 * library https://github.com/fernandezpablo85/scribe-java (collectively, "Third
 * Party Code"). Microsoft Mobile is not the original author of the Third Party
 * Code.
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package jsimple.oauth.builder.api;

import jsimple.oauth.extractors.*;
import jsimple.oauth.model.OAuthConfig;
import jsimple.oauth.model.Token;
import jsimple.oauth.model.Verb;
import jsimple.oauth.oauth.OAuth10aServiceImpl;
import jsimple.oauth.oauth.OAuthService;
import jsimple.oauth.services.HMACSha1SignatureService;
import jsimple.oauth.services.SignatureService;
import jsimple.oauth.services.TimestampService;
import jsimple.oauth.services.TimestampServiceImpl;

/**
 * Default implementation of the OAuth protocol, version 1.0a
 * <p/>
 * This class is meant to be extended by concrete implementations of the API, providing the endpoints and
 * endpoint-http-verbs.
 * <p/>
 * If your Api adheres to the 1.0a protocol correctly, you just need to extend this class and define the getters for
 * your endpoints.
 * <p/>
 * If your Api does something a bit different, you can override the different extractors or services, in order to
 * fine-tune the process. Please read the javadocs of the interfaces to get an idea of what to do.
 *
 * @author Pablo Fernandez
 */
public abstract class DefaultOAuthApi10a implements OAuthApi {
    /**
     * Returns the access token extractor.
     *
     * @return access token extractor
     */
    public AccessTokenExtractor getAccessTokenExtractor() {
        return new TokenExtractorImpl();
    }

    /**
     * Returns the base string extractor.
     *
     * @return base string extractor
     */
    public BaseStringExtractor getBaseStringExtractor() {
        return new BaseStringExtractorImpl();
    }

    /**
     * Returns the header extractor.
     *
     * @return header extractor
     */
    public HeaderExtractor getHeaderExtractor() {
        return new HeaderExtractorImpl();
    }

    /**
     * Returns the request token extractor.
     *
     * @return request token extractor
     */
    public RequestTokenExtractor getRequestTokenExtractor() {
        return new TokenExtractorImpl();
    }

    /**
     * Returns the signature service.
     *
     * @return signature service
     */
    public SignatureService getSignatureService() {
        return new HMACSha1SignatureService();
    }

    /**
     * Returns the timestamp service.
     *
     * @return timestamp service
     */
    public TimestampService getTimestampService() {
        return new TimestampServiceImpl();
    }

    /**
     * Returns the method for the access token endpoint (defaults to POST)
     *
     * @return access token endpoint verb
     */
    public String getAccessTokenVerb() {
        return Verb.POST;
    }

    /**
     * Returns the verb for the request token endpoint (defaults to POST)
     *
     * @return request token endpoint verb
     */
    public String getRequestTokenVerb() {
        return Verb.POST;
    }

    /**
     * Returns the URL that receives the request token requests.
     *
     * @return request token URL
     */
    public abstract String getRequestTokenEndpoint();

    /**
     * Returns the URL that receives the access token requests.
     *
     * @return access token URL
     */
    public abstract String getAccessTokenEndpoint();

    /**
     * Returns the URL where you should redirect your users to authenticate your application.
     *
     * @param requestToken the request token you need to authorize
     * @return the URL where you should redirect your users
     */
    public abstract String getAuthorizationUrl(Token requestToken);

    /**
     * Returns the {@link OAuthService} for this Api
     *
     * @param config config info (apiKey, apiSecret, callback, scope)
     */
    public OAuthService createService(OAuthConfig config) {
        return new OAuth10aServiceImpl(this, config);
    }
}

File: jsimple-oauth/c#/jsimple/oauth/builder/api/DefaultOAuthApi10a.cs
namespace jsimple.oauth.builder.api {

    using jsimple.oauth.extractors;
    using OAuthConfig = jsimple.oauth.model.OAuthConfig;
    using Token = jsimple.oauth.model.Token;
    using Verb = jsimple.oauth.model.Verb;
    using OAuth10aServiceImpl = jsimple.oauth.oauth.OAuth10aServiceImpl;
    using OAuthService = jsimple.oauth.oauth.OAuthService;
    using HMACSha1SignatureService = jsimple.oauth.services.HMACSha1SignatureService;
    using SignatureService = jsimple.oauth.services.SignatureService;
    using TimestampService = jsimple.oauth.services.TimestampService;
    using TimestampServiceImpl = jsimple.oauth.services.TimestampServiceImpl;

    /// <summary>
    /// Default implementation of the OAuth protocol, version 1.0a
    /// <p/>
    /// This class is meant to be extended by concrete implementations of the API, providing the endpoints and
    /// endpoint-http-verbs.
    /// <p/>
    /// If your Api adheres to the 1.0a protocol correctly, you just need to extend this class and define the getters for
    /// your endpoints.
    /// <p/>
    /// If your Api does something a bit different, you can override the different extractors or services, in order to
    /// fine-tune the process. Please read the javadocs of the interfaces to get an idea of what to do.
    /// 
    /// @author Pablo Fernandez
    /// </summary>
    public abstract class DefaultOAuthApi10a : OAuthApi {
        /// <summary>
        /// Returns the access token extractor.
        /// </summary>
        /// <returns> access token extractor </returns>
        public virtual AccessTokenExtractor AccessTokenExtractor {
            get {
                return new TokenExtractorImpl();
            }
        }

        /// <summary>
        /// Returns the base string extractor.
        /// </summary>
        /// <returns> base string extractor </returns>
        public virtual BaseStringExtractor BaseStringExtractor {
            get {
                return new BaseStringExtractorImpl();
            }
        }

        /// <summary>
        /// Returns the header extractor.
        /// </summary>
        /// <returns> header extractor </returns>
        public virtual HeaderExtractor HeaderExtractor {
            get {
                return new HeaderExtractorImpl();
            }
        }

        /// <summary>
        /// Returns the request token extractor.
        /// </summary>
        /// <returns> request token extractor </returns>
        public virtual RequestTokenExtractor RequestTokenExtractor {
            get {
                return new TokenExtractorImpl();
            }
        }

        /// <summary>
        /// Returns the signature service.
        /// </summary>
        /// <returns> signature service </returns>
        public virtual SignatureService SignatureService {
            get {
                return new HMACSha1SignatureService();
            }
        }

        /// <summary>
        /// Returns the timestamp service.
        /// </summary>
        /// <returns> timestamp service </returns>
        public virtual TimestampService TimestampService {
            get {
                return new TimestampServiceImpl();
            }
        }

        /// <summary>
        /// Returns the method for the access token endpoint (defaults to POST)
        /// </summary>
        /// <returns> access token endpoint verb </returns>
        public virtual string AccessTokenVerb {
            get {
                return Verb.POST;
            }
        }

        /// <summary>
        /// Returns the verb for the request token endpoint (defaults to POST)
        /// </summary>
        /// <returns> request token endpoint verb </returns>
        public virtual string RequestTokenVerb {
            get {
                return Verb.POST;
            }
        }

        /// <summary>
        /// Returns the URL that receives the request token requests.
        /// </summary>
        /// <returns> request token URL </returns>
        public abstract string RequestTokenEndpoint {get;}

        /// <summary>
        /// Returns the URL that receives the access token requests.
        /// </summary>
        /// <returns> access token URL </returns>
        public abstract string AccessTokenEndpoint {get;}

        /// <summary>
        /// Returns the URL where you should redirect your users to authenticate your application.
        /// </summary>
        /// <param name="requestToken"> the request token you need to authorize </param>
        /// <returns> the URL where you should redirect your users </returns>
        public abstract string getAuthorizationUrl(Token requestToken);

        /// <summary>
        /// Returns the <seealso cref="OAuthService"/> for this Api
        /// </summary>
        /// <param name="config"> config info (apiKey, apiSecret, callback, scope) </param>
        public virtual OAuthService createService(OAuthConfig config) {
            return new OAuth10aServiceImpl(this, config);
        }
    }

}